### PR TITLE
Add validation to dynamic header table update

### PR DESCRIPTION
**Pull request by cjyclaire**
addressing #146 
https://http2.github.io/http2-spec/compression.html#encoding.context.update

### CODE REVIEW

**Comment by cjyclaire** [comment]
Hmmm might need to rework on this, I need to make sure the checking with the limit set from : SETTINGS_HEADER_TABLE_SIZE

**Comment by cjyclaire** [comment]
might add more docs later, review ready

File: lib/http/2/compressor.rb [hunk 1]
```diff
@@ -103,6 +103,7 @@ def initialize(**options)
         @table = []
         @options = default_options.merge(options)
         @limit = @options[:table_size]
+        @dynamic_table_size_update_limit = options[:settings_header_table_size] || 4096
```
Comment (igrigorik) [inline lib/http/2/compressor.rb]: Why do we need this if @limit is already tracking this?

File: lib/http/2/compressor.rb [hunk 2]
```diff
@@ -146,6 +147,9 @@ def process(cmd)
 
         case cmd[:type]
         when :changetablesize
+          if cmd[:value] > @dynamic_table_size_update_limit
```
Comment (igrigorik) [inline lib/http/2/compressor.rb]: Same as above, why can't we use @limit?

File: lib/http/2/compressor.rb [hunk 3]
```diff
@@ -470,6 +474,10 @@ def initialize(**options)
         @cc = EncodingContext.new(options)
       end
 
+      def dynamic_table_size_update_limit=(size)
```
Comment (igrigorik) [inline lib/http/2/compressor.rb]: Is this necessary if we're using @limit?

File: lib/http/2/compressor.rb (hunk 1 above)
Comment (cjyclaire) [reply lib/http/2/compressor.rb]: my bad, I confused myself originally, then I realized it was set properly for the limit already

**Comment by cjyclaire** [comment]
@igrigorik  Thanks for the review! I simplified the diff and made sure rubycop passed, mind take a second look?  : D (through jruby is failing still but not related to the changes I made ...)

Apologies, I confused my self originally that why I made it complicated first, I read the spec section again and double check the code base, @limit does satisfied the need already.

File: lib/http/2/compressor.rb [hunk 4]
```diff
@@ -146,6 +146,9 @@ def process(cmd)
 
         case cmd[:type]
         when :changetablesize
+          if cmd[:value] > @limit
+            fail ProtocolError, 'dynamic table size update exceed limit'
```
Comment (igrigorik) [inline lib/http/2/compressor.rb]: > A value that exceeds this limit MUST be treated as a decoding error.

WDYT of raising [CompressionError](https://github.com/igrigorik/http-2/blob/master/lib/http/2/error.rb#L16-L19)?

Comment (cjyclaire) [reply lib/http/2/compressor.rb]: Sounds good, updated in the latest commit : )

**Comment by ioquatix** [comment]
I ran into this implementation while working through h2spec and my fork of this code in `protocol-hpack`, and found that this implementation seems wrong to me.

The `@limit` is related to the table size, but it's adjusted to be smaller if table size is changed. This doesn't seem right, in my implementation I'm changing `@limit` => `@table_size` and adding `@table_size_limit` which is set independently.

Just FYI.

**Comment by ioquatix** [comment]
I've been playing around with this a bit more, in order to try and be compliant with RFC, but I don't completely understand how we should handle settings relating to table size.

>    SETTINGS_HEADER_TABLE_SIZE (0x1):  Allows the sender to inform the
>       remote endpoint of the maximum size of the header compression
>       table used to decode header blocks, in octets.  The encoder can
>       select any size equal to or less than this value by using
>       signaling specific to the header compression format inside a
>       header block (see [COMPRESSION]).  The initial value is 4,096
>       octets.

So, if we receive this on the server, when we encode headers (e.g. for response headers), we should limit the dynamic table to the specified size?

**Comment by ioquatix** [comment]
Okay, so I found a potential problem with the implementation.

Essentially, if you change the table size, you need to emit the right `:change_table_size` command. But right now the implementation doesn't appear to do it. So, I implemented it.

https://github.com/socketry/protocol-hpack/commit/51879fbea2ea205aba3fed59072a94acaca04d49

It was pretty straight forward to add once I understood how it was supposed to be used. My implementation is pretty naive, it just changes the table size to the maximum the client allows. I assume this is resource intensive.

**Pull request merged**